Futures order position tracking & FTX scaled collateral calculation

gloriousCode opened the pull request:
# PR Description
This PR is based on the following branch, minus any backtester changes [here](https://github.com/gloriousCode/gocryptotrader/tree/future-order-handling). 

As getting futures support working for the backtester requires many changes, I thought I'd extract some reviewable parts of my work. This PR does the following:
- Adds a new futures position controller, responsible for tracking futures positions across all exchanges
- Adds the ability to track futures positions across multiple orders (see "things to consider" below)
  - Tracks orders from position opening to closure, allowing for positions to be amended or having the position direction swapped
- Calculates PNL based on orders in a position
- Calculates collateral for FTX
- Adds new GRPC command `getfuturesposition` (note: mocking my sweet PNL is banned)

```
gctcli> go run . --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="BTC-1231"
{
 "totalOrders": 22,
 "totalRealisedPNL": "0.07871883",
 "totalPNL": "0.07871883",
 "positions": [
  {
   "status": "CLOSED",
   "realisedPNL": "0.11357933",
   "openingDate": "2021-10-31 23:38:26 +0000",
   "closingDate": "2021-11-01 02:25:43 +0000"
  },
  {
   "status": "CLOSED",
   "realisedPNL": "-0.00130672",
   "openingDate": "2021-12-02 01:44:41 +0000",
   "closingDate": "2021-12-02 01:45:32 +0000"
  },
  {
   "status": "CLOSED",
   "realisedPNL": "-0.01402298",
   "openingDate": "2021-12-06 01:25:32 +0000",
   "closingDate": "2021-12-06 01:28:35 +0000"
  },
  {
   "status": "CLOSED",
   "realisedPNL": "-0.02453374",
   "openingDate": "2021-12-06 02:18:02 +0000",
   "closingDate": "2021-12-06 02:20:41 +0000"
  },
  {
   "status": "CLOSED",
   "realisedPNL": "-0.02332539",
   "openingDate": "2021-12-06 04:13:28 +0000",
   "closingDate": "2021-12-06 04:17:35 +0000"
  },
  {
   "status": "CLOSED",
   "realisedPNL": "-0.01530603",
   "openingDate": "2021-12-06 04:29:48 +0000",
   "closingDate": "2021-12-06 04:31:09 +0000"
  },
  {
   "status": "CLOSED",
   "realisedPNL": "0.04363436",
   "openingDate": "2021-12-20 22:54:29 +0000",
   "closingDate": "2021-12-21 02:37:12 +0000"
  }
 ]
}
```
- Adds new GRPC command `getcollateral`
```
gctcli> go run . getcollateral ftx futures false true
{
 "totalCollateral": "1337.5468421889777368103900077",
 "currencyBreakdown": [
  {
   "currency": "USD",
   "scaledCollateral": "1333.404998391084"
  },
  {
   "currency": "USDT",
   "scaledCollateral": "0.000000009508303729750508",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "BTC",
   "scaledCollateral": "4.141841689076325",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "AUD",
   "scaledCollateral": "0.0000000020176057803824997",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "BNB",
   "scaledCollateral": "0.000002097291502300257",
   "scaledToCurrency": "USD"
  }
 ]
}
```

These features have been implemented for FTX as that's what I'm currently targeting for backtesting futures work

## Things to consider
- Futures orders are much more complicated to track how much money is being made. Long running and complicated futures positions won't be calculated properly and will be a future PR.
  - At the moment, some of the PNL calculation is done based on _opening price_, however, complex orders which append to a short or long after a position is already open have different prices to calculate PNL from. This is not yet happening and is solely on the opening price.
 -  For PNL calculation to work accurately, you must track from your opening position! Tracking midway while you have positions open will cause problems as there will be no endpoint/an inaccurate endpoint. If this happens to you, use the optional paramter `overwrite` when requesting position data via GRPC. It will clear any existing data for the exchange+currencypair+asset


## Type of change

- [x] New feature (non-breaking change which adds functionality)


## How has this been tested
I've added tests to all new functions as well as added some additional coverage where it came up.

- [x] go test ./... -race
- [x] golangci-lint run
- [x] Test coverage of all new additions

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [x] Any dependent changes have been merged and published in downstream modules

shazbert reviewed (changes requested):
This is quite a loaded PR, appreciate you pulled it out from the main work you are doing. Looks really good so far, Still need to do more tests after 3 coffees. 🚀

gloriousCode commented on cmd/documentation/engine_templates/order_manager.tmpl:
```diff
@@ -4,6 +4,7 @@
 + The order manager subsystem stores and monitors all orders from enabled exchanges with API keys and `authenticatedSupport` enabled
 + It can be enabled or disabled via runtime command `-ordermanager=false` and defaults to true
 + All orders placed via GoCryptoTrader will be added to the order manager store
++ Any futures based order will be tracked via the [futures positions controller](/exchanges/orders/README.md) which can be used to track PNL. Use GRPC command [getfuturesposition](https://api.gocryptotrader.app/#gocryptotrader_getfuturesposition) to view position data for an exchange, asset, pair
```
It used to... I must not have committed it 😢

gloriousCode commented on cmd/gctcli/commands.go:
```diff
@@ -4723,17 +4723,309 @@ func findMissingSavedCandleIntervals(c *cli.Context) error {
+	if !strings.EqualFold(status, "any") &&
+		!strings.EqualFold(status, "open") &&
+		!strings.EqualFold(status, "closed") &&
+		status != "" {
```
Why?

shazbert replied:
I guess if it was unset by accident?

gloriousCode replied:
The check is there to prevent people from seeing the help description of ` "limit return to position statuses - open, closed, any"`, entering `lol` and expecting a response

shazbert replied:
Ahh okay, I get you. I just assumed an empty string was invalid. 👍

gloriousCode replied:
Oh, with the way github renders, I thought you were asking to remove the entire `if` statement, not just the `""` check. I went with an empty string assuming "any" and made it valid.

It also doesn't look like I do any filtering on the RPCServer side, so I'll add that

gloriousCode commented on engine/order_manager_test.go:
```diff
@@ -1152,3 +1153,211 @@ func Test_getActiveOrders(t *testing.T) {
+		t.Errorf("received '%v', expected '%v'", err, order.ErrPositionsNotLoadedForExchange)
+	}
+	o = nil
+	_, err = o.GetFuturesPositionsForExchange("test", asset.Futures, cp)
```
The function that `GetFuturesPositionsForExchange` calls (`GetPositionsForExchange`) is fully covered and checked, so I didn't think it was necessary to retest it. But this sounds like something I'd ask of you 😄 so I can do that

gloriousCode commented on engine/order_manager_types.go:
```diff
@@ -37,11 +41,12 @@ type orderManagerConfig struct {
-	commsManager    iCommsManager
-	exchangeManager iExchangeManager
-	wg              *sync.WaitGroup
+	m                         sync.RWMutex
```
I've added this ez trello ticket [here](https://trello.com/c/8G54M6Ei/141-investigate-and-change-order-manager-mutex-from-rwmutex-to-mutex). I think it could derail things in this PR, but its a good point you brought up

shazbert replied:
Thank you sir!

gloriousCode commented on exchanges/ftx/ftx_wrapper.go:
```diff
@@ -1254,3 +1266,161 @@ func (f *FTX) GetAvailableTransferChains(ctx context.Context, cryptocurrency cur
+		collateral, err := f.ScaleCollateral(ctx, &collateralAssets[i])
+		if err != nil {
+			if errors.Is(err, errCollateralCurrencyNotFound) {
+				log.Error(log.ExchangeSys, err)
```
👍 I'll update it in `ScaleCollateral`

gloriousCode commented on exchanges/ftx/ftx_types.go:
```diff
@@ -880,3 +880,14 @@ type StakeReward struct {
 }
+
+// CollateralWeightHolder stores collateral weights over the lifecycle of the application
+type CollateralWeightHolder map[string]CollateralWeight
```
For such a simple struct, this change would increase memory allocations as it needs to put every entry into the heap. Performance is similar enough that I don't think this change is necessary

shazbert replied:
Interesting you are getting allocations? I am not seeing that from tests I have done. Diff below:

```diff
--- a/exchanges/ftx/ftx.go
+++ b/exchanges/ftx/ftx.go
@@ -1479,7 +1479,7 @@ func (f *FTX) FetchExchangeLimits(ctx context.Context) ([]order.MinMaxLevel, err
 // LoadCollateralWeightings sets the collateral weights for
 // currencies supported by FTX
 func (f *FTX) LoadCollateralWeightings(ctx context.Context) error {
-       f.collateralWeight = make(map[string]CollateralWeight)
+       f.collateralWeight = make(map[string]*CollateralWeight)
        // taken from https://help.ftx.com/hc/en-us/articles/360031149632-Non-USD-Collateral
        // sets default, then uses the latest from FTX
        f.collateralWeight.load("1INCH", 0.9, 0.85, 0.0005)
@@ -1642,25 +1642,23 @@ func (c CollateralWeightHolder) isLoaded() bool {
 func (c CollateralWeightHolder) loadTotal(code string, weighting float64) {
        currencyCollateral, ok := c[code]
        if !ok {
-               currencyCollateral = CollateralWeight{Total: weighting}
+               currencyCollateral = &CollateralWeight{Total: weighting}
        } else {
                currencyCollateral.Total = weighting
        }
-       c[code] = currencyCollateral
 }

 func (c CollateralWeightHolder) loadInitialMarginFraction(code string, imf float64) {
        currencyCollateral, ok := c[code]
        if !ok {
-               currencyCollateral = CollateralWeight{InitialMarginFractionFactor: imf}
+               currencyCollateral = &CollateralWeight{InitialMarginFractionFactor: imf}
        } else {
                currencyCollateral.InitialMarginFractionFactor = imf
        }
-       c[code] = currencyCollateral
 }

 func (c CollateralWeightHolder) load(code string, initial, total, imfFactor float64) {
-       c[code] = CollateralWeight{
+       c[code] = &CollateralWeight{
                Initial:                     initial,
                Total:                       total,
                InitialMarginFractionFactor: imfFactor,
diff --git a/exchanges/ftx/ftx_test.go b/exchanges/ftx/ftx_test.go
index a76db9060..bd575ae98 100644
--- a/exchanges/ftx/ftx_test.go
+++ b/exchanges/ftx/ftx_test.go
@@ -1969,3 +1969,16 @@ func TestLoadCollateralWeightings(t *testing.T) {
                t.Fatal("expected some weight")
        }
 }
+
+// 33623899            36.01 ns/op            0 B/op          0 allocs/op // No pointer
+// 58665273            19.72 ns/op            0 B/op          0 allocs/op // With pointer
+func BenchmarkCollateral(b *testing.B) {
+       err := f.LoadCollateralWeightings(context.Background())
+       if err != nil {
+               b.Fatal(err)
+       }
+
+       for i := 0; i < b.N; i++ {
+               f.collateralWeight.loadInitialMarginFraction(currency.BTC.String(), 0.1)
+       }
+}
```

shazbert replied:
Didn't test load function and it does have alloc. Doesn't periodically update either so inconsequential. Resolving.

gloriousCode commented on exchanges/order/futures.go:
```diff
@@ -0,0 +1,606 @@
+		realisedPNL = realisedPNL.Add(pnlHistory[i].RealisedPNLBeforeFees)
+		totalFees = totalFees.Add(pnlHistory[i].Fee)
+	}
+	if realisedPNL.IsZero() {
+		return decimal.Zero
+	}
```
🎉  I think I can get rid of this now. I used to calculate realisedPNL every iteration which caused issues

shazbert commented on exchanges/ftx/ftx_wrapper.go:
```diff
@@ -1254,3 +1266,171 @@ func (f *FTX) GetAvailableTransferChains(ctx context.Context, cryptocurrency cur
+		if !ok {
+			return decimal.Zero, fmt.Errorf("%s %s %w", f.Name, calc.CollateralCurrency, errCollateralCurrencyNotFound)
+		}
+		if calc.CollateralAmount.IsPositive() {
```
NOTE: This is not a change request but we might be able to get a better value with something like this included. 
```go
if calc.USDPrice.Equal(decimal.Zero) {
			last, err := ticker.FindLast(currency.NewPair(calc.CollateralCurrency, currency.USD), calc.Asset)
			if err != nil {
				log.Warn(log.Global, "not found bro")
			}
			calc.USDPrice = decimal.NewFromFloat(last)
		}

```

Also when doing online I am coming across zero value for SRM_LOCKED which is expected but other tokens which are zero.

gloriousCode replied:
This was a good bug catch. Offline calculation wasn't sending USD values. I think I'll put it at the rcpserver level. 

Does "offline calculation" as a command term seem accurate to you? By offline, I mean that when scaling collateral, it already has all the data it needs (which will be used in the backtester for example), but for this RPCServer function, I'll need to retreive USD value data per currency.

Also, what are you asking of me for the online part?

shazbert reviewed (changes requested):
Thanks for addressing all the change requests. 

Position tracking testing:

```json
go run ./... --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="BTC-perp"
{
 "totalOrders": 4,
 "totalRealisedPNL": "-0.2770445625",
 "positions": [
  {
   "status": "CLOSED",
   "unrealisedPNL": "0",
   "realisedPNL": "-0.1481740275",
   "openingDate": "2021-10-25 23:05:28 +0000",
   "closingDate": "2021-10-25 23:06:48 +0000"
  },
  {
   "status": "CLOSED",
   "unrealisedPNL": "0",
   "realisedPNL": "-0.128870535",
   "openingDate": "2021-10-25 23:20:30 +0000",
   "closingDate": "2021-10-25 23:20:37 +0000"
  }
 ]
}
```

Can confirm position return calc from front end.
![image](https://user-images.githubusercontent.com/9391715/149039149-d2f8044d-c49b-476f-8510-64d702d1161f.png)

Would be good in future to see a percentage increase or decrease as well. 

Collateral calculation testing:

attempt 1 offline:
```json
go run ./... --timeout=5m getcollateral --exchange="ftx" --asset="spot" --i --c
{
 "totalCollateral": "1.07902616",
 "currencyBreakdown": [
  {
   "currency": "USD",
   "scaledCollateral": "1.07902616"
  }
 ]
}
```

another attempt same account:

```json 
go run ./... --timeout=5m getcollateral --exchange="ftx" --asset="spot" --i --c
{
 "totalCollateral": "4007.3620206",
 "currencyBreakdown": [
  {
   "currency": "BTC",
   "scaledCollateral": "0",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "OKB",
   "scaledCollateral": "0",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "BNB",
   "scaledCollateral": "0",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "USD",
   "scaledCollateral": "4007.3620206"
  },
  {
   "currency": "XRP",
   "scaledCollateral": "0",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "SRM",
   "scaledCollateral": "0",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "FTT",
   "scaledCollateral": "0",
   "scaledToCurrency": "USD"
  }
 ]
}
```

Don't know why it takes a couple requests to populate it all just yet. 

Also: When we designate a sub-account via config we cannot utilize getcollateral for main or other sub-accounts, which for now is not a big issue.

gloriousCode commented:
> Don't know why it takes a couple requests to populate it all just yet.

From our discussions, it may have been a result of you using a subaccount, but this isn't something I have been able to replicate. If it comes up again, please provide steps to recreate. The RPCServer now returns a subaccount name on both `GetCollateral` and `GetFuturesPosition` results if it is set/specified

```
 go run . --timeout=5m getcollateral ftx futures false true
{
 "subAccount": "helloMoto",
 "totalCollateral": "0"
}
```
or 
```
 go run . --timeout=5m getcollateral --exchange="ftx" --asset="futures" --subaccount="helloMoto" --i --c
{
 "subAccount": "helloMoto",
 "totalCollateral": "0"
}
```

shazbert commented on exchanges/ftx/ftx_wrapper.go:
```diff
@@ -1254,3 +1266,179 @@ func (f *FTX) GetAvailableTransferChains(ctx context.Context, cryptocurrency cur
+}
+
+// CalculateTotalCollateral scales collateral and determines how much collateral you can use for positions
+func (f *FTX) CalculateTotalCollateral(ctx context.Context, collateralAssets []order.CollateralCalculator) (*order.TotalCollateralResponse, error) {
```
Sub-accounts are mixing up with main account funds. We are calling 
[DEBUG]|REQUESTER| 14/01/2022 09:31:38 |FTX attempt 1 request path: https://ftx.com/api/wallet/balances multiple times and 
[DEBUG]|REQUESTER| 14/01/2022 09:31:38 |FTX attempt 1 request path: https://ftx.com/api/wallet/coins when I send the RPC request. 

Seems like there is no differentiation between balances and accounts and in the outgoing headers there is nothing to differentiate as well.

gloriousCode replied:
I've added basic subaccount overriding in FTX now, this allows subaccount querying for retreiving collateral when calculating online.

I've also minimised the number of calls

shazbert reviewed (changes requested):
Thanks for those changes, just one thing I found when getting collateral. And a pedantic suggestion. 👨‍🎤

shazbert commented on engine/rpcserver.go:
```diff
@@ -4106,3 +4109,241 @@ func (s *RPCServer) CurrencyStateTradingPair(_ context.Context, r *gctrpc.Curren
+			}
+		}
+	} else if len(ai.Accounts) > 0 {
+		acc = &ai.Accounts[0]
```
This is causing an account mismatch. Also out of scope issue with hold values when util exch.FetchAccountInfo(ctx, a) above as discussed in PM.

shazbert reviewed (changes requested):
Thanks, one issue left over. Then looks good to go.

shazbert reviewed (approved):
Tested ACK. GetCollarateral via RPC works as expected and the change with the account balances I have cross-referenced with the front end for available amounts. LGTM.  Thanks for making those changes. 🤝 🚀

thrasher- reviewed (changes requested):
This is quite a cool feature! Upon testing, there appears to be a few issues:

On a new account when testing for an open order unrealised pnl calculation, it detects the correct open date but shows its number of orders increasing each time I request it plus the unrealised pnl is 0 when infact the front-end shows a positive number. No modifications to the position have been done since opening:

```
$ ./gctcli --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="MATIC-PERP" --start="2021-01-01 00:00:00"
{
 "totalOrders": 160,
 "subAccount": "test",
 "positions": [
  {
   "status": "OPEN",
   "currentDirection": "LONG",
   "unrealisedPNL": "0",
   "realisedPNL": "0",
   "openingDate": "2022-01-24 02:36:05 +0000"
  }
 ]
}

...

./gctcli --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="MATIC-PERP" --start="2021-01-01 00:00:00"
{
 "totalOrders": 192,
 "subAccount": "test",
 "positions": [
  {
   "status": "OPEN",
   "currentDirection": "LONG",
   "unrealisedPNL": "0",
   "realisedPNL": "0",
   "openingDate": "2022-01-24 02:36:05 +0000"
  }
 ]
}

...

./gctcli --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="MATIC-PERP" --start="2021-01-01 00:00:00"
{
 "totalOrders": 224,
 "subAccount": "test",
 "positions": [
  {
   "status": "OPEN",
   "currentDirection": "LONG",
   "unrealisedPNL": "0",
   "realisedPNL": "0",
   "openingDate": "2022-01-24 02:36:05 +0000"
  }
 ]
}
```
![image](https://user-images.githubusercontent.com/4685270/151297112-c078a4d5-1661-43ab-8f5f-969a80afbcd8.png)

It looks like the USDT weight is out on this when comparing to the front-end, it would be good to show those and the original balance:

```
$ ./gctcli getcollateral --exchange=ftx --asset=futures --includebreakdown --subaccount=test
{
 "subAccount": "test",
 "totalCollateral": "98288.0851368795625",
 "currencyBreakdown": [
  {
   "currency": "USD",
   "scaledCollateral": "748.1033118795625",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "USDT",
   "scaledCollateral": "97539.981825",
   "scaledToCurrency": "USD"
  }
 ]
}
```

Front-end:

![image](https://user-images.githubusercontent.com/4685270/151299189-6b27d54d-14d1-4193-be83-0983aeadcfdb.png)

Another thing that I've noticed is that staked coins are showing up in the collateral list when they should be skipped instead as FTX does not include them.

gloriousCode commented:
> This is quite a cool feature! Upon testing, there appears to be a few issues:
> 
> On a new account when testing for an open order unrealised pnl calculation, it detects the correct open date but shows its number of orders increasing each time I request it 

Great find on that bug, I was appending orders to the wrong location 👀 

>  plus the unrealised pnl is 0 when infact the front-end shows a positive number. No modifications to the position have been done since opening:
>PNL $365.94

I've updated it to now retrieve the latest ticker price if a position is open and calculate the unrealised PNL from that... I found it strange that I did not do this initially 👀 

> It looks like the USDT weight is out on this when comparing to the front-end,

You found a nice bug in how I was loading scaling! This should be more accurate now

> it would be good to show those and the original balance:

There is some weirdness from FTX's side, but I've added the original balance to the response and filtered out some data. Hopefully this presents more accurately for your account and I'll work with you soon to verify
```
{
 "totalCollateral": "1337.19004354644379564530073",
 "currencyBreakdown": [
  {
   "currency": "USD",
   "originalAmount": "1333.55079036"
  },
  {
   "currency": "USDT",
   "scaledCollateral": "0.00000000950911196464203",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "BTC",
   "originalAmount": "0.0001",
   "scaledCollateral": "3.6392513275208995",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "AUD",
   "scaledCollateral": "0.0000000019782378627165",
   "scaledToCurrency": "USD"
  },
  {
   "currency": "BNB",
   "scaledCollateral": "0.0000018429875463179422",
   "scaledToCurrency": "USD"
  }
 ]
}
```

shazbert commented on exchanges/ftx/ftx_test.go:
```diff
@@ -2019,11 +2039,11 @@ func TestLoadCollateralWeight(t *testing.T) {
-	if cw.Total != 2 {
-		t.Errorf("expected '2', received '%v'", cw.InitialMarginFractionFactor)
+	if cw.Initial != 2 {
+		t.Errorf("expected '2', received '%v'", cw.Total)
```
Total -> Initial

shazbert commented on exchanges/ftx/ftx_test.go:
```diff
@@ -2019,11 +2039,11 @@ func TestLoadCollateralWeight(t *testing.T) {
-	if cw.Initial != 1 {
-		t.Errorf("expected '1', received '%v'", cw.Total)
+	if cw.Total != 1 {
+		t.Errorf("expected '1', received '%v'", cw.InitialMarginFractionFactor)
```
cw.InitialMarginFractionFactor -> cw.Total

shazbert commented on exchanges/ftx/ftx_wrapper.go:
```diff
@@ -1403,11 +1407,16 @@ func (f *FTX) CalculateTotalCollateral(ctx context.Context, subAccount string, c
-					})
+					breakDown := order.CollateralByCurrency{
+						Currency: collateralAssets[x].CollateralCurrency,
+
```
rm ln

shazbert commented on exchanges/order/futures.go:
```diff
@@ -91,6 +91,50 @@ func (c *PositionController) GetPositionsForExchange(exch string, item asset.Ite
+// using the latest ticker data
+func (c *PositionController) UpdateOpenPositionUnrealisedPNL(exch string, item asset.Item, pair currency.Pair, last float64, updated time.Time) (decimal.Decimal, error) {
+	if c == nil {
+		return decimal.Zero, common.ErrNilPointer
```
fmt.Errorf("position controller %w", common.ErrNilPointer) ?

shazbert commented on exchanges/order/futures.go:
```diff
@@ -91,6 +91,50 @@ func (c *PositionController) GetPositionsForExchange(exch string, item asset.Ite
+	}
+	c.m.Lock()
+	defer c.m.Unlock()
+	if !item.IsFutures() {
+		return decimal.Zero, fmt.Errorf("%v %v %v %w", exch, item, pair, ErrNotFuturesAsset)
+	}
```
can we move this check above the locks?

shazbert commented on exchanges/order/futures_test.go:
```diff
@@ -729,3 +744,61 @@ func TestTrackPNLByTime(t *testing.T) {
+	t.Parallel()
+	pc := SetupPositionController()
+
+	_, err := pc.UpdateOpenPositionUnrealisedPNL("hi", asset.Futures, currency.NewPair(currency.BTC, currency.USDT), 2, time.Now())
```
👋

shazbert reviewed (changes requested):
Thanks for new changes, tested collateral and positional functionality. 

```
go run ./... --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="BTC-perp"
2022/01/31 12:20:26 rpc error: code = Unknown desc = ftx futures BTCperp no positions loaded for exchange
exit status 1
```
This was on a new test account, suggesting adding an additional field to the error for sub account on error.

Also I was wondering if getcollateral flag --asset might not be needed, some collateral is spot holdings. (Don't need to change this though, just bringing it up).

gloriousCode commented:
> This was on a new test account, suggesting adding an additional field to the error for sub account on error.

I've added it, but since that function doesn't actually have a subaccount parameter, its going off your own supplied config. I would hope that a user knows what credentials they are using. Though a thought for the future would be to add a new command to get some account details like subaccount/whether auth is endpoints are allowed or not.
```
go run . --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="eth-0325"  --verbose=true
... ftx futures eth0325 no positions loaded for exchange for subaccount: gloriousCodeLamboMoonFund
```

> Also I was wondering if getcollateral flag --asset might not be needed, some collateral is spot holdings.

For FTX, yes, but Binance has different collateral based on coin-margined futures and usdt-margined futures. I can remove it from the philosophy of coding for right now, if you would like, but it would be re-added later.

shazbert commented:
> > This was on a new test account, suggesting adding an additional field to the error for sub account on error.
> 
> I've added it, but since that function doesn't actually have a subaccount parameter, its going off your own supplied config. I would hope that a user knows what credentials they are using. Though a thought for the future would be to add a new command to get some account details like subaccount/whether auth is endpoints are allowed or not.
> 
> ```
> go run . --timeout=5m getfuturesposition --exchange="ftx" --asset="futures" --pair="eth-0325"  --verbose=true
> ... ftx futures eth0325 no positions loaded for exchange for subaccount: gloriousCodeLamboMoonFund
> ```
> 
> > Also I was wondering if getcollateral flag --asset might not be needed, some collateral is spot holdings.
> 
> For FTX, yes, but Binance has different collateral based on coin-margined futures and usdt-margined futures. I can remove it from the philosophy of coding for right now, if you would like, but it would be re-added later.

Thanks sir, and happy to keep it.

shazbert reviewed (approved):
tACK! Thanks for making those changes.

thrasher- reviewed (changes requested):
Thank you for making those changes! I can see that the number of trades matches the front-end and the unrealised pnl is on par with the front-end:

```
{
 "totalOrders": 32,
 "subAccount": "test",
 "totalUnrealisedPNL": "8162.7048",
 "positions": [
  {
   "status": "OPEN",
   "currentDirection": "LONG",
   "unrealisedPNL": "8162.7048",
   "realisedPNL": "0",
   "openingDate": "2022-01-24 02:36:05 +0000"
  }
 ]
}
```
![image](https://user-images.githubusercontent.com/4685270/151908734-9420af9f-513b-48ef-933b-93d4cd327df9.png)

Some notes when testing:

- Collateral scaling for coins which have a specific amount staked/locked looks to use the full balance rather than the available free amount for the collateral calcs
- If a user has a negative balance contribution (either UPNL or borrowed amounts), it would be nice to show that since that can impact the collateral available for futures positions

gloriousCode commented:
I believe I have addressed the issues raised. There is a much more detailed breakdown for online and offline calculations of how collateral is allocated, which is a big improvement to the original PR! 🎉 The below numbers have been tweaked
Online:
```
go run . --timeout=5m getcollateral --exchange="ftx" --asset="futures" --i
{
 "collateral_currency": "USD",
 "total_collateral": "1337.11934794326187602124435",
 "maintenance_collateral": "1337.11934794326187602124435",
 "free_collateral": "7.25119994000309988797435",
 "locked_collateral": "1329.86814800325877613327",
 "locked_breakdown": {
  "locked_in_spot_margin_funding_offers": "1329.86814800325877613327"
 },
 "currency_breakdown": [
  {
   "currency": "USD",
   "original_amount": "1337.11934794",
   "scaled_total": "1337.11934794326187602124435",
   "scaled_maintenance": "1337.11934794326187602124435",
   "scaled_free": "7.25119994000309988797435",
   "scaled_locked": "1329.86814800325877613327",
   "locked_breakdown": {
    "locked_in_spot_margin_funding_offers": "1329.86814800325877613327"
   }
  },
 ]
}
```
Offline:
```
go run . --timeout=5m getcollateral --exchange="ftx" --asset="futures" --i --c
{
 "total_collateral": "1337.11934794",
 "maintenance_collateral": "1337.11934794",
 "free_collateral": "7.25119994",
 "locked_collateral": "1329.868148",
 "currency_breakdown": [
  {
   "currency": "USD",
   "original_amount": "1337.11934794",
   "scaled_total": "1337.11934794",
   "scaled_maintenance": "1337.11934794",
   "scaled_free": "7.25119994",
   "scaled_locked": "1329.868148"
  },
 ]
}
```
In order to do this, I needed to update the values stored via `/exchanges/account` struct `Balance`, which I've updated across exchanges. See below:
```
.\gctcli.exe getaccountinfo ftx futures
{
 "exchange": "FTX",
 "accounts": [
  {
   "id": "main",
   "currencies": [
    {
     "currency": "USD",
     "total_value": 1337.11934794,
     "hold": 1333.868148,
     "free": 9.05112978,
     "freeWithoutBorrow": 3.25119994
    },
}
```

Hopefully your more adventurous subaccounts are also accurate now
# 🤞

shazbert commented on exchanges/ftx/ftx_wrapper.go:
```diff
@@ -1348,104 +1365,233 @@ func (f *FTX) ScaleCollateral(ctx context.Context, subAccount string, calc *orde
+			calc.CollateralAssets[i].UnrealisedPNL = decimal.Zero
+			for i := range pos {
+				var futurePair currency.Pair
+				futurePair, err = currency.NewPairFromString(pos[i].Future)
```
Suggestion: Could change the Future field type from String to currency.Pair

shazbert commented on exchanges/ftx/ftx_wrapper.go:
```diff
@@ -1348,104 +1365,233 @@ func (f *FTX) ScaleCollateral(ctx context.Context, subAccount string, calc *orde
+				breakDown := order.CollateralByCurrency{
+					Currency:          calc.CollateralAssets[x].CollateralCurrency,
+					ScaledTotal:       total.Mul(scaling),
+					ScaledMaintenance: total.Mul(scaling),
```
This is the same as above?

shazbert commented on exchanges/ftx/ftx_wrapper.go:
```diff
@@ -1348,104 +1365,233 @@ func (f *FTX) ScaleCollateral(ctx context.Context, subAccount string, calc *orde
+					}
+
+					breakDown.ScaledLockedBreakdown = &order.CollateralLockedBreakdown{
+						LockedInStakes:                  lockedS.Mul(scaling),
```
these `lockedS.Mul(scaling)` and others can be taken out of this scope to be used below in results locked breakdown.

shazbert commented on exchanges/order/futures_types.go:
```diff
@@ -54,26 +56,46 @@ type PNLCalculation interface {
+	Error                 error
+}
+
+type CollateralLockedBreakdown struct {
```
COmment please

shazbert reviewed (changes requested):
Epic changes, just minor things I have found, while testing I received some sweet info back.

shazbert reviewed (approved):
Tick Tack front and back.

gloriousCode commented:
With the latest changes, I've increased accuracy by using FTX's collateral endpoints for online calculation.
```
go run . --timeout=5m getcollateral --exchange="ftx" --asset="futures" --i
{
 "collateral_currency": "USD",
 "maintenance_collateral": "629.3432900776618",
 "available_collateral": "620.5548580776618",
 "used_collateral": "13.96590313844",
 "used_breakdown": {
  "locked_in_stakes": "5.06030313844",
  "locked_in_spot_margin_funding_offers": "5",
  "used_in_futures": "3.9056"
 },
 "currency_breakdown": [
  {
   "currency": "USD",
   "original_amount": "623.08392936",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "1",
   "scaled_total": "623.08382936",
   "scaled_free": "618.08382936",
   "scaled_used": "5",
   "used_breakdown": {
    "locked_in_spot_margin_funding_offers": "5"
   },
   "weighting": "1"
  },
  {
   "currency": "FTT",
   "original_amount": "0.14464336",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "41.852",
   "scaled_total": "5.06030313844",
   "scaled_free": "0",
   "scaled_used": "5.06030313844",
   "used_breakdown": {
    "locked_in_stakes": "5.06030313844"
   },
   "weighting": "0"
  },
  {
   "currency": "ETH",
   "original_amount": "0.0009998",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "2714.03735037",
   "scaled_total": "2.5778198157549297",
   "scaled_free": "2.5778198157549297",
   "scaled_used": "0",
   "weighting": "0.95"
  },
  {
   "currency": "BTC",
   "original_amount": "0.0001",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "38962.14258366",
   "scaled_total": "3.79880890190685",
   "scaled_free": "3.79880890190685",
   "scaled_used": "0",
   "weighting": "0.975"
  }
 ],
 "position_breakdown": [
  {
   "currency": "BTC-0325",
   "size": "0.0001",
   "open_order_size": "0",
   "position_size": "0.0001",
   "mark_price": "39056",
   "required_margin": "1",
   "total_collateral_used": "3.9056"
  }
 ]
}
```

Offline calculation does have an issue for FTT where its collateral is dependent on a user setting and so will include it regardless, where online does not. I haven't wanted to hard-code FTT/SRM/SRM_LOCKED into this PR, but if you want me to I can look into it
```
go run . --timeout=5m getcollateral --exchange="ftx" --asset="futures" --i --c
{
 "available_collateral": "625.1956703435795",
 "used_collateral": "9.8299161888285",
 "currency_breakdown": [
  {
   "currency": "USD",
   "original_amount": "623.07922936",
   "approx_fair_market_value": "0",
   "scaled_total": "623.07922936",
   "scaled_free": "618.07922936",
   "scaled_used": "5",
   "weighting": "1"
  },
  {
   "currency": "FTT",
   "original_amount": "0.14464336",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "42.049",
   "scaled_total": "5.778003212408",
   "scaled_free": "0.9480870235795",
   "scaled_used": "4.8299161888285",
   "weighting": "0.95"
  },
  {
   "currency": "ETH",
   "original_amount": "0.0009998",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "2728",
   "scaled_total": "2.45470896",
   "scaled_free": "2.45470896",
   "scaled_used": "0",
   "weighting": "0.9"
  },
  {
   "currency": "BTC",
   "original_amount": "0.0001",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "39091",
   "scaled_total": "3.713645",
   "scaled_free": "3.713645",
   "scaled_used": "0",
   "weighting": "0.95"
  }
 ]
}
```

A quick note, the offline calculation method is designed for the Backtester and its' provided data, so it does not require the same level of detail at present. However, the above snippet demonstrates that the local calculation functions accurately scale the data it does utilise

shazbert reviewed (changes requested):
Nothing much from me with the new collateral changes. In the `used_breakdown`, `locked in stakes` seems to be scaling to USD for SRM locked and some others.  
```
 {
   "currency": "SRM_LOCKED",
   "original_amount": "48.18570376",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "1.86",
   "scaled_total": "89.6199711348",
   "scaled_free": "0",
   "scaled_used": "89.6199711348",
   "used_breakdown": {
    "locked_in_stakes": "89.6199711348"
   },
   "weighting": "0"
  },
```

While other used breakdowns seem fine:

```
 {
   "currency": "BNB",
   "original_amount": "2.31845413",
   "scaled_to_currency": "USD",
   "approx_fair_market_value": "353.38338378",
   "scaled_total": "4.10118554975956283",
   "scaled_free": "1.78805754975956283",
   "scaled_used": "2.313128",
   "used_breakdown": {
    "locked_in_spot_margin_funding_offers": "2.1974716"
   },
   "weighting": "0.95"
  },
```

Good stuff!

gloriousCode commented:
Okay, I think I've done it _this_ time 😄 I swear! I've made it more clear and the valuations are more accurate
```
{
 "collateral_currency": "USD",
 "maintenance_collateral": "629.3638590701802 USD",
 "available_collateral": "625.6802840701802 USD",
 "used_collateral": "3.7975 USD",
 "used_breakdown": {
  "used_in_futures": "3.7975 USD"
 },
 "currency_breakdown": [
  {
   "currency": "USD",
   "total_funds": "623.3059901 USD",
   "available_for_use_as_collateral": "623.3061901 USD",
   "approx_fair_market_value": "1 USD",
   "weighting": "1",
   "collateral_contribution": "623.3061901 USD",
   "scaled_to_currency": "USD"
  },
  {
   "currency": "FTT",
   "excluded_from_collateral": true,
   "total_funds": "0.14464336 FTT",
   "available_for_use_as_collateral": "0.02373389 FTT",
   "approx_fair_market_value": "41.371 USD",
   "weighting": "0",
   "collateral_contribution": "0 USD",
   "scaled_to_currency": "USD",
   "used_breakdown": {
    "locked_in_stakes": "5.00214568337 USD"
   }
  },
  {
   "currency": "ETH",
   "total_funds": "0.0009998 ETH",
   "available_for_use_as_collateral": "0.0009998 ETH",
   "approx_fair_market_value": "2622.77754812 USD",
   "weighting": "0.95",
   "collateral_contribution": "2.4911403429798572 USD",
   "scaled_to_currency": "USD"
  },
  {
   "currency": "BTC",
   "total_funds": "0.0001 BTC",
   "available_for_use_as_collateral": "0.0001 BTC",
   "approx_fair_market_value": "37748.24233026 USD",
   "weighting": "0.975",
   "collateral_contribution": "3.68045362720035 USD",
   "scaled_to_currency": "USD"
  }
 ],
 "position_breakdown": [
  {
   "currency": "BTC-0325",
   "size": "0.0001",
   "open_order_size": "0",
   "position_size": "0.0001",
   "mark_price": "37975 USD",
   "required_margin": "1",
   "total_collateral_used": "3.7975 USD"
  }
 ]
}
```

It also works with SRM_LOCKED:
```
 {
   "currency": "SRM_LOCKED",
   "excluded_from_collateral": true,
   "total_funds": "48.1644741 SRM_LOCKED",
   "available_for_use_as_collateral": "0.00292357 SRM_LOCKED",
   "approx_fair_market_value": "1.927 USD",
   "weighting": "0",
   "collateral_contribution": "0 USD",
   "scaled_to_currency": "USD",
   "used_breakdown": {
    "locked_in_stakes": "92.80730787131 USD"
   }
  },
```

shazbert reviewed (approved):
Collateral looks good and the added lovely display currency is a nice touch! Good stuff! tACK.

thrasher- reviewed (changes requested):
I can see that futures UPNL and online collateral calculations are matching up with the front-end 🎉 We're still missing the "additional collateral used" breackdown field per spot margin borrowed item, but that's something I have a ticket with FTX open ATM.

![image](https://user-images.githubusercontent.com/4685270/155641820-311a8aea-361e-4017-81c3-bc889efd441c.png)

The locked_in_stakes value shows the serum amount in USD, but should be SRM amount * current SRM market price

shazbert reviewed (approved):
Recent changes look good and the collateral with the front end matches for me. Online compared to offline is a $1 difference in avail collateral 👍.

thrasher- commented on cmd/gctcli/commands.go:
```diff
@@ -4723,17 +4723,325 @@ func findMissingSavedCandleIntervals(c *cli.Context) error {
+	var overwrite bool
+	if c.IsSet("overwrite") {
+		overwrite = c.Bool("overwrite")
+	} else if c.Args().Get(2) != "" {
```
```suggestion
	} else if c.Args().Get(8) != "" {
```

thrasher- commented on cmd/gctcli/commands.go:
```diff
@@ -4723,17 +4723,325 @@ func findMissingSavedCandleIntervals(c *cli.Context) error {
+	var verbose bool
+	if c.IsSet("verbose") {
+		verbose = c.Bool("verbose")
+	} else if c.Args().Get(6) != "" {
```
Do you mean 7?

thrasher- commented on cmd/gctcli/commands.go:
```diff
@@ -4723,17 +4723,325 @@ func findMissingSavedCandleIntervals(c *cli.Context) error {
+	}
+
+	var subAccount string
+	if c.IsSet("subaccount") {
```
Subaccount is last in the ArgsUsage blurb, but second last here

thrasher- reviewed (changes requested):
Thanks for making those additional changes! Collateral matches FTX front-end and also same with the futures PNL on an account with staking, margin borrowing and free assets 🎉 Only have some minor nits left

thrasher- reviewed (approved):
tACK, nice work!

shazbert reviewed (changes requested):
This is pretty much a tACK; just one thing I noticed:

Offline calculations have approx_fair_market_value as 0 for USD. 
```
{
   "currency": "USD",
   "total_funds": "4018.00442214 USD",
   "available_for_use_as_collateral": "0.03206414 USD",
   "approx_fair_market_value": "0 USD",
   "weighting": "1",
   "collateral_contribution": "0.03206414 USD",
   "scaled_to_currency": "USD",
   "funds_in_use": "4017.972358 USD"
  },
```

shazbert reviewed (approved):
ChangesACK.
